skglm/covariance.py

```diff
@@ -0,0 +1,218 @@
+                    Theta[indices != col, col] = theta_12
+                    Theta[col, indices != col] = theta_12
+                    Theta[col, col] = (1/s_22 +
+                                       theta_12 @
+                                       inv_Theta_11 @
+                                       theta_12)
+                    theta_22 = Theta[col, col]
+
+                    # Updating W
+                    W[col, col] = (1/(theta_22 -
+                                      theta_12 @
+                                      inv_Theta_11 @
+                                      theta_12))
+                    w_22 = W[col, col]
+
+                    w_12 = (-w_22 * inv_Theta_11 @ theta_12)
+                    W[indices != col, col] = w_12
+                    W[col, indices != col] = w_12
+
+                    # Maybe W_11 can be done smarter ?
+                    W[_11] = (inv_Theta_11 +
+                              np.outer(w_12,
+                                       w_12)/w_22)
+
+            if np.linalg.norm(Theta - Theta_old) < self.tol:
+                if self.verbose:
+                    print(f"Weighted Glasso converged at CD epoch {it + 1}")
+                break
+        else:
+            if self.verbose:
+                print(
+                    f"Not converged at epoch {it + 1}, "
+                    f"diff={np.linalg.norm(Theta - Theta_old):.2e}"
+                )
+        self.precision_, self.covariance_ = Theta, W
+        self.n_iter_ = it + 1
+
+        return self
+
+
+class AdaptiveGraphicalLasso():
+    """ An adaptive version of the Graphical Lasso that solves non-convex penalty
+    variations using the reweighting strategy from Candès et al., 2007."""
+
+    def __init__(
+        self,
+        alpha=1.,
+        strategy="log",
+        n_reweights=5,
+        max_iter=1000,
+        tol=1e-8,
+        warm_start=False,
+    ):
+        self.alpha = alpha
+        self.strategy = strategy
+        self.n_reweights = n_reweights
+        self.max_iter = max_iter
+        self.tol = tol
+        self.warm_start = warm_start
+
+    def fit(self, S):
+        glasso = GraphicalLasso(
+            alpha=self.alpha,
+            algo="primal",
+            max_iter=self.max_iter,
+            tol=self.tol,
+            warm_start=True)
+        Weights = np.ones(S.shape)
+        self.n_iter_ = []
+        for it in range(self.n_reweights):
+            glasso.weights = Weights
+            glasso.fit(S)
+            Theta = glasso.precision_
+            Weights = update_weights(Theta, self.alpha, strategy=self.strategy)
+            self.n_iter_.append(glasso.n_iter_)
+            # TODO print losses for original problem?
+            glasso.covariance_ = np.linalg.pinv(Theta, hermitian=True)
+        self.precision_ = glasso.precision_
+        self.covariance_ = glasso.covariance_
+        return self
+
+
+def update_weights(Theta, alpha, strategy="log"):
+    if strategy == "log":
```

Comment:
Can you try to directly use penalty.derivative as is done in IterativeReweightedL1 ?